keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1207,27 +1207,89 @@ Details to consider:
 
 #### CEL Expression Composition
 
+Expression Composition is a technique to define a set of variables each from
+a given expression and allow validation expressions to refer the variables.
+
 ##### Use Cases
 
+###### Reusing/memorizing an expensive computation and lazy evaluation
+
+For a CEL expression that takes a significant time to evaluate, especially
+these that cost O(N^2) time or worse, it would be nice to only run it once
+and only when necessary. If multiple validation expressions used the same
+expression, that expression could be refactored out into a variable.
+Because of the evaluation of a composited variable is lazy, if the value of some variable
+does not affect the result of an expression due to boolean short-circuit evaluation,
+the variable can be omitted and does not incur runtime cost.
+
 ###### Code re-use for complicated expressions
 
 A CEL expression may not be computationally expensive, but could still be
 intricate enough that copy-pasting could prove to be a bad decision later on
-in time. With the addition of the `messageExpression` field, more copy-pasting
-is expected as well. If a sufficiently complex expression ended up copy-pasted everywhere,
-and then needs to be updated somehow, it will need that update in every place
-it was copy-pasted. A variable, on the other hand, will only need to be updated
-in one place.
+in time. For example, a sub-expression is likely used both in the validation and
+to format `messageExpression`. If a sufficiently complex expression ended up
+copy-pasted everywhere, and then needs to be updated somehow, it will need that update in every place
+it was copy-pasted. A variable, on the other hand, will only need to be updated in one place.
+
+##### Variables
+
+Each CEL "program" is a single expression. There is no support for variable
+assignment. This can result in redundant code to traverse maps/arrays or
+dereference particular fields.
+
+We can support this in much the same way as cel-policy-template `terms`. These
+can be lazily evaluated when they are resolved during the evaluation of
+the main expression (cel-policy-template does this).
+
+The policy spec now has an additional `variables` section. This is an array
+containing one or more `name` and `expression` pairs, which can be used/re-used by
+the policy's validation expressions. These results are memoized on a
+per-validation basis, so if multiple expressions use the same spec variables,
+the expression that calculates the variable's value will only run once.
+
+The variables can be accessed as members of `variables`, which is an object
+that is exposed to CEL expressions (both validation expressions and other variables).
+
+For example:
+
+```yaml
+  variables:
+    - name: metadataList
+      expression: "spec.list.map(x, x.metadata)"
+    - name: itemMetadataNames
+      expression: "variables.metadataList.map(m, m.name)"
```

Comment:
This looks right to me.

Variable names are keys right and must be unique right?  I only bring this up because I've been dealing with https://github.com/kubernetes/kubernetes/issues/118261 today, which has problems caused by pod container env variables being reassigned (e.g. `A='x'`, `A=${A}y`).

Reply:
Added requirement for unique identifier names.